src/general/block-endpoint-edit-role/src.rego

```diff
@@ -0,0 +1,32 @@
+package blockendpointeditrole
+
+violation[{"msg": msg}] {
+    input.review.object.metadata.name == "system:aggregate-to-edit"
+    endpointRule(input.review.object.rules[_])
+    msg := "ClusterRole system:aggregate-to-edit should not allowed endpoint permissions"
+}
+
+violation[{"msg": msg}] {
+    input.review.object.metadata.name == "system:aggregate-to-edit"
+    not disableAutoupdate(input.review.object.metadata.annotations)
+    msg := "ClusterRole system:aggregate-to-edit is required rbac.authorization.kubernetes.io/autoupdate=false after reconciled none endpoint permission"
+}
+
+endpointRule(rule) {
+    "endpoints" == rule.resources[_]
+    hasEditVerb(rule.verbs)
+}
+
+hasEditVerb(verbs) {
+	"create" == verbs[_]
+}
+hasEditVerb(verbs) {
+    "patch" == verbs[_]
+}
+hasEditVerb(verbs) {
+	"update" == verbs[_]
+}
+
+disableAutoupdate(annotations) {
+	annotations["rbac.authorization.kubernetes.io/autoupdate"] == "false"
```

Comment:
Do we want to require disabling auto update?

It sounds like in the future, disabling auto update will not be necessary. If the resource gets auto-updated to a violating form, either:

- the auto update will be rejected
- the violation will show up in audit

Advising users that they should disable auto update as part of a violation message definitely seems like a good idea.

Reply:
I removed auto update violation function and put the advise in the violation message.